text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1546 @@
+  and unknown at compilation time. These would include the scalable vector types
+  mentioned in the motivation below, or those that implement `Sized` but not
+  `const Sized` in the RFC.
+- The bounds on the generic parameters of a function may be referred to simply
+  as the bounds on the function (e.g. "the caller's bounds").
+
+## Acknowledgements
+[acknowledgements]: #acknowledgements
+
+This RFC wouldn't have been possible without the reviews and feedback of
+[@JamieCunliffe][author_jamiecunliffe], [@JacobBramley][ack_jacobbramley],
+[@nikomatsakis][author_nikomatsakis] and [@scottmcm][ack_scottmcm];
+[@eddyb][ack_eddyb] for the `externref` future possibility; the expertise
+of [@compiler-errors][ack_compiler_errors] on the type system and suggesting
+the use of const traits; [@fee1-dead][ack_fee1dead] for reviewing the usage
+of const traits and fixing the ASCII diagrams; and the authors of all of the
+prior art for influencing these ideas.
+
+# Motivation
+[motivation]: #motivation
+
+Introducing a hierarchy of `Sized` and elaborating on the implications of constness
+on `Sized` traits resolves blockers for other RFCs which have had significant interest.
+
+## Runtime-sized types and scalable vector types
+[runtime-sized-types-and-scalable-vector-types]: #runtime-sized-types-and-scalable-vector-types
+
+Rust already supports [SIMD][rfc_simd] (*Single Instruction Multiple Data*),
+which allows operating on multiple values in a single instruction. Processors
+have SIMD registers of a known, fixed length and a variety of intrinsics
+which operate on these registers. For example, x86-64 introduced 128-bit SIMD
+registers with SSE, 256-bit SIMD registers with AVX, and 512-bit SIMD registers
+with AVX-512, and Arm introduced 128-bit SIMD registers with Neon.
+
+As an alternative to releasing SIMD extensions with greater bit widths, Arm and
+RISC-V have vector extensions (SVE/Scalable Vector Extension and the "V" Vector
+Extension/RVV respectively) where the bit width of vector registers depends on the
+CPU implementation, and the instructions which operate these registers are bit
+width-agnostic.
+
+As a consequence, these types are not `Sized` in the Rust sense, as the size of
+a scalable vector cannot be known during compilation, but is a runtime constant.
+For example, the size of these types could be determined by inspecting the value
+in a register - this is not available at compilation time and the value may
+differ between any given CPU implementation. Both SVE and RVV have mechanisms to
+change the system's vector length (up to the maximum supported by the CPU
+implementations) but this is not supported by the proposed ABI for these types.
+
+However, despite not implementing `Sized`, these are value types which should
+implement `Copy` and can be returned from functions, can be variables on the
+stack, etc. These types should implement `Copy` but given that `Sized` is a
```

Comment:
How does stack allocation works for the "runtime constant" types, at binary level? Is it different from dynamic stack allocation?
Does it work for SVE specifically, or it's possible to do for other "runtime constant" types too?

I assume that the "runtime constant" value becomes available somewhere before the call to `main`?
Or earlier (e.g. during linking), and you need to specify the "runtime env" during the build?

Reply:
> How does stack allocation work for the "runtime constant" types, at binary level? Is it different from dynamic stack allocation?
> Does it work for SVE specifically, or it's possible to do for other "runtime constant" types too?

~~I believe it's just "dynamic `alloca`", I don't recall ever seeing anything in LLVM about "runtime constant" stack frame sizes.~~
(while searching for examples to add elsewhere in this comment I've found that while LLVM `alloca` *is* used, [it's more "typed" than I expected](https://github.com/search?q=repo%3Allvm%2Fllvm-project+%2Fstack-id%3A+scalable-vector%7CTargetStackID%3A%3AScalableVector%2F&type=code), and the stack frame layout is aware of the distinction - cc @nikic, not sure how this will look once `alloca Type` is gone)

<sub>While it would be neat to treat such values (and those derived from them with simple arithmetic) as "relocation-time constants" and have the dynamic linker patch them into e.g. as many instruction immediates as possible, [I have recently argued against the desirability of storing vector registers to memory](https://github.com/rust-lang/rfcs/pull/3268#issuecomment-2468886230) (outside of e.g. context switching and other low-level tasks), so I'm skeptical of the value of relocation-level support.</sub>

---

> I assume that the "runtime constant" value becomes available somewhere before the call to `main`?

In theory, some mechanism like ELF auxv could be used, but AIUI it's just an instruction reading a "special register":
- ARM has `mrs`/`msr` instructions to move between general-purpose and special registers
  - SVE also has [`rdvl`](https://developer.arm.com/documentation/ddi0602/2024-09/SVE-Instructions/RDVL--Read-multiple-of-vector-register-size-to-scalar-register-) and similarly helpful instructions
  - combined with `alloca`, [Clang/LLVM-produced asm gets wild](https://github.com/llvm/llvm-project/blob/e477989a055f92f6ca63fc8f76929cde81d33e44/llvm/test/CodeGen/AArch64/sve-alloca.ll) (Ctrl+F `vl`)
- RISC-V more specifically calls them "CSR"s ("Control and Status Register"s)
  - the relevant `size_of` value for RVV, `vlenb`, is a read-only CSR
  - (while RVV has other CSRs and instructions, see below why they're not relevant for `size_of`)

I'd compare it to using x86 `cpuid` to e.g. determine the SIMD extension with the largest registers (and it similarly is "not supposed to change during execution" AFAICT).

LLVM seems to abstract this concept in its [`@llvm.vscale.iN`](https://llvm.org/docs/LangRef.html#llvm-vscale-intrinsic) intrinsic, documented as:
> `vscale` is a positive value that is constant throughout program execution, but is unknown at compile time.
  - anything more fine-grained than `vscale` has to go either through target-specific intrinsics *or* the newer [predicated vector intrinsics](https://llvm.org/docs/LangRef.html#int-vp), which take *both* a mask *and* an "EVL" ("Effective Vector Length") and in theory could (eventually) be useful for a `core::simd`-like abstraction

---

There's also Linux-specific syscalls/documentation:
- SVE: https://docs.kernel.org/arch/arm64/sve.html
  - `prctl(PR_SVE_GET_VL)` as alternative to getting VL through instructions
    - [not widely used AFAICT](https://github.com/search?q=prctl%28PR_SVE_GET_VL%29+-path%3A%2Fstress%7Cvalidation%7Ctest%7C%28%5C.%28rst%7Cmd%7Ctxt%7Chtml%29%24%29%2F&type=code) (outside the Java HotSpot VM, for JIT reasons I assume) and [Clang/LLVM definitely use `rdvl`](https://github.com/llvm/llvm-project/blob/e477989a055f92f6ca63fc8f76929cde81d33e44/llvm/test/CodeGen/AArch64/sve-vscale.ll#L8-L15) and [related instructions](https://github.com/llvm/llvm-project/blob/e477989a055f92f6ca63fc8f76929cde81d33e44/llvm/test/CodeGen/AArch64/sve-alloca.ll)
  - can also set current VL (or defer to `execve`, i.e. set it for a child process)
  - not expecting changing VL on the fly to be compatible with code using SVE C intrinsics
- RVV: https://docs.kernel.org/arch/riscv/vector.html
  - less control than SVE, as RVV only has these two lengths:
    - `vlenb`: the *hardwired* µarch implementation maximum (which is what non-`const` `size_of` would supposedly return, [much to my surprise](https://github.com/rust-lang/rfcs/pull/3268#issuecomment-2468886230))
    - `vl`: the dynamic length, meant to be set by intrinsics and auto-vectorized loops *on the fly* (to remove the need for any scalar loops, as e.g. the last iteration of the vector loop can still work with `vl` as low as `1`)
  - > To get the availability of V in an ELF program, please read `COMPAT_HWCAP_ISA_V` bit of `ELF_HWCAP` in the auxiliary vector.
    - so there is something available through ELF auxv, but it's just extension bits (and ARM has them too)

Out of curiosity, I've also looked at the ELF auxv HWCAP for x86 and AFAICT it's [`cpuid(eax=1).edx`](https://github.com/torvalds/linux/blob/9f16d5e6f220661f73b36a4be1b21575651d8833/arch/x86/include/asm/elf.h#L241-L254) which [sadly only goes up to SSE2 in terms of extension bits](https://en.wikipedia.org/wiki/CPUID#EAX.3D1:_Processor_Info_and_Feature_Bits) (newer ones being in ECX), and [HWCAP2 is a couple of kernel-controlled features](https://github.com/torvalds/linux/blob/9f16d5e6f220661f73b36a4be1b21575651d8833/arch/x86/include/uapi/asm/hwcap2.h).

---

Also, while looking up the LLVM-specific examples to link above, I came across SME/SVE2 "streaming mode" [which is a whole new can of worms](https://llvm.org/docs/AArch64SME.html), and I hope doesn't require even stranger types.

Reply:
> (while searching for examples to add elsewhere in this comment I've found that while LLVM `alloca` _is_ used, [it's more "typed" than I expected](https://github.com/search?q=repo%3Allvm%2Fllvm-project+%2Fstack-id%3A+scalable-vector%7CTargetStackID%3A%3AScalableVector%2F&type=code), and the stack frame layout is aware of the distinction - cc @nikic, not sure how this will look once `alloca Type` is gone)

I'd expect that `alloca` would accept a [`TypeSize`](https://llvm.org/doxygen/classllvm_1_1TypeSize.html), which is either a fixed size or a size multiplied by vscale. I think that would cover it?

Reply:
This thread has made me much more sceptical that these types provide good motivation for this feature. It's interesting that `externref` style types have similar requirements: they need to be `Copy`, they can be passed as arguments, returned as results, used as local variables, etc, but importantly they aren't `Sized` (or even `ValueSized`)... Similarly, references to both types are weird because they aren't pointers (or at least shouldn't be in the case of vector types as I think @eddyb is suggesting).

I haven't yet come up with a productive solution though...

Reply:
> Similarly, references to both types are weird because they aren't pointers (or at least shouldn't be in the case of vector types as I think @eddyb is suggesting).

As I understand it, scalable vectors can be used behind pointers, only `externref` style types cannot.